teps/0138-decouple-api-and-feature-versioning.md

```diff
@@ -72,8 +91,331 @@ Per [TEP-0033](https://github.com/tektoncd/community/blob/main/teps/0033-tekton-
+| stable |	true  |	yes	| no  |
+| stable |	false |	no	| no  |
+
+For new beta features(e.g. matrix in the future):
+
+| enable-api-fields |	legacy-enable-beta-features-by-default | enabled in `v1beta1`? | enabled in `v1`? |
+| ------ | ------ | --- | --- |
+| beta   |	true  |	yes |	yes |
+| beta   |	false |	yes |	yes |
+| stable |	true  |	no  |	no  |
+| stable |	false |	no  |	no  |
+
+Once all existing beta features become stable, `legacy-enable-beta-features-by-default` can be removed and we will deprecate and then remove `legacy-enable-beta-features-by-default` and to use `stable` `enable-api-fields`. We would default `true` for the new flag and after 9 months, we default `enable-api-fields` to `stable` to preserve the existing behavior.
+
+#### Pros
+- This will provide a smoother transition for switching default `enable-api-fields` value back to `stable`.
+
+#### Cons
+- This is a backwards incompatible change.
+- It is not clear what to do with alpha features for the feature promotion process when there are two `enable-api-fields` related flags, which might lead to confusions.
+- This is adding complexity to both the implementations and the testing matrix for the newly introduced flag.
+
+
+### Make validation changes only for new beta features
+This alternative proposes keeping beta features on by default. It will keep the current implementations and validations for beta features in `v1` and `v1beta1` apiVersion and only make new features have the synchronized validations across different apiVersions. This also proposes to keep the default value of `enable-api-fields` as `beta`,
+
+#### Pros
+- This is backwards compatible.
+- Changes are minimal for the current codebase with probably only documentation required.
+#### Cons
+- We would still need to use `beta` as the default stability of features that are turned on. Some beta features are still coupled in feature and API versioning. This will not allow cluster operators to opt-in only stable features in `v1beta1`.
+
+### Give 9-month warning before breaking changes and default to stable
+This alternative proposes the validation change for stable features in `v1beta1` to be done with giving a notice for the breaking change for 9 months.
+
+#### Cons:
+- Since `v1beta1` only has 12 months of support period left, making a breaking change to it might have more impacts on `v1beta1` users than its worth after 9 months.
+
+## Implementation Plan
+
+For each new feature driven by api fields, a new feature flag using FeatureFlag struct will be added to the existent set of feature flags. This struct will include the stability of feature and whether it is turned on or locked to default. Currently it will only work for new features without `enable-api-fields`, see future work for more details of the option to introduce `enable-api-fields=none` to use per feature flags for new features.
+
+
+```go
+type FeatureFlag struct {
+    // Name of the feature flag
+    Name string
```

Comment:
What is the name used for? Perhaps for logging purposes? 
Perhaps we could do something similar to what we have for behavioural flags:

```
type APIFeatureFlags struct {
   Foo  FeatureFlag
   Bar   FeatureFlag
}
```

Reply:
?

Reply:
I agree the name field could be kept for logging purposes.
As for the grouping of all the API feature flags, my understanding is that such struct would make sense if we would like to turn on behavioural features of a certain stability level. But I am not sure if this is going to be necessary while we fully switch to per-feature flags (i.e. after all current beta/alpha features stabilized)?